package/src/core/with-plugins.ts

```diff
@@ -0,0 +1,57 @@
+import type { AstroIntegration } from "astro";
+import type { NonEmptyArray } from "../internal/types.js";
+import type {
+	AnyPlugin,
+	ExtendedHooks,
+	HookParameters,
+	Hooks,
+} from "./types.js";
+
+/**
+ * Allows to extend hooks with custom parameters. Advanced.
+ *
+ * @param {object} params
+ * @param {string} params.name The integration name
+ * @param {Array<AnyPlugin>} params.plugins
+ * @param {import("astro".AstroIntegration["hooks"])} params.hooks
+ */
+export const withPlugins = <TPlugins extends NonEmptyArray<AnyPlugin>>({
+	name,
+	plugins,
+	hooks: providedHooks,
+}: { name: string; plugins: TPlugins; hooks: ExtendedHooks<TPlugins> }) => {
+	const resolvedPlugins = Object.values(
+		(() => {
+			const _plugins: Record<string, AnyPlugin> = {};
+			for (const plugin of plugins) {
+				_plugins[plugin.name] = plugin;
+			}
+			return _plugins;
+		})(),
+	);
```

Comment:
```suggestion
const resolvedPlugins = Object.values(Object.fromEntries(plugins.map(plugin => [plugin.name, plugin])))
```

This is a bit hard to read, it could be simplified to something like this

Also, not sure if this is intentional, but this will cause plugins that share names but have different hooks to overwrite each other

Reply:
It's intentional and that's waaaay better than you!

Reply:
It's up to the plugin author to provide a name that is safe enough. Types and and plugins will be overwritten based on the name. I think this limitation is fine for most usecases (remember that in most cases a utility is enough) but let me know what you think!

Reply:
I agree, the current implementation is fine for most use-cases. This is not a common pattern and most likely won't cause any problems, it just feels wrong that plugin names have to be unique across all hooks when they can only be used inside one hook. I am ok with keeping this I just wanted to point it out as a limitation I haven't seen mentioned anywhere. The docs mention this:

> Plugins support overrides. That means that if 2 plugins declare ***the same name and hook***, the latest will be kept

Which suggests that the name is unique to a single hook instead of all hooks

Reply:
Yes I updated it to match the upcoming behavior!

Reply:
That being said, I'm happy to make docs clearer if you have suggestions